core/src/main/java/org/infinispan/marshall/core/internal/PrimitiveExternalizer.java

```diff
@@ -0,0 +1,683 @@
+         for (double v : obj) out.writeDouble(v);
+      } else {
+         out.writeByte(ARRAY_LARGE);
+         out.writeInt(len);
+         for (double v : obj) out.writeDouble(v);
+      }
+   }
+
+   private void writeFloatArray(float[] obj, ObjectOutput out) throws IOException {
+      final int len = obj.length;
+      if (len == 0) {
+         out.writeByte(ARRAY_EMPTY);
+      } else if (len <= 256) {
+         out.writeByte(ARRAY_SMALL);
+         out.writeByte(len);
+         for (float v : obj) out.writeFloat(v);
+      } else if (len <= 65536) {
+         out.writeByte(ARRAY_MEDIUM);
+         out.writeShort(len);
+         for (float v : obj) out.writeFloat(v);
+      } else {
+         out.writeByte(ARRAY_LARGE);
+         out.writeInt(len);
+         for (float v : obj) out.writeFloat(v);
+      }
+   }
+
+   private void writeIntArray(int[] obj, ObjectOutput out) throws IOException {
+      final int len = obj.length;
+      if (len == 0) {
+         out.writeByte(ARRAY_EMPTY);
+      } else if (len <= 256) {
+         out.writeByte(ARRAY_SMALL);
+         out.writeByte(len);
+         for (int v : obj) out.writeInt(v);
+      } else if (len <= 65536) {
+         out.writeByte(ARRAY_MEDIUM);
+         out.writeShort(len);
+         for (int v : obj) out.writeInt(v);
+      } else {
+         out.writeByte(ARRAY_LARGE);
+         out.writeInt(len);
+         for (int v : obj) out.writeInt(v);
+      }
+   }
+
+   private void writeLongArray(long[] obj, ObjectOutput out) throws IOException {
+      final int len = obj.length;
+      if (len == 0) {
+         out.writeByte(ARRAY_EMPTY);
+      } else if (len <= 256) {
+         out.writeByte(ARRAY_SMALL);
+         out.writeByte(len);
+         for (long v : obj) out.writeLong(v);
+      } else if (len <= 65536) {
+         out.writeByte(ARRAY_MEDIUM);
+         out.writeShort(len);
+         for (long v : obj) out.writeLong(v);
+      } else {
+         out.writeByte(ARRAY_LARGE);
+         out.writeInt(len);
+         for (long v : obj) out.writeLong(v);
+      }
+   }
+
+   private void writeShortArray(short[] obj, ObjectOutput out) throws IOException {
+      final int len = obj.length;
+      if (len == 0) {
+         out.writeByte(ARRAY_EMPTY);
+      } else if (len <= 256) {
+         out.writeByte(ARRAY_SMALL);
+         out.writeByte(len);
+         for (short v : obj) out.writeShort(v);
+      } else if (len <= 65536) {
+         out.writeByte(ARRAY_MEDIUM);
+         out.writeShort(len);
+         for (short v : obj) out.writeShort(v);
+      } else {
+         out.writeByte(ARRAY_LARGE);
+         out.writeInt(len);
+         for (short v : obj) out.writeShort(v);
+      }
+   }
+
+   private void writeObjectArray(Object[] obj, ObjectOutput out) throws IOException {
+      final int len = obj.length;
+      if (len == 0) {
+         out.writeByte(ARRAY_EMPTY);
+      } else if (len <= 256) {
+         out.writeByte(ARRAY_SMALL);
+         out.writeByte(len);
```

Comment:
If you wrote `len - 1` instead, then you wouldn't need a branch to support length 256 when reading.